platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java

```diff
@@ -266,67 +286,99 @@ public void initInputDevices() {
-			if (device != null) {
-				int sources = device.getSources();
-				if (((sources & InputDevice.SOURCE_GAMEPAD) == InputDevice.SOURCE_GAMEPAD) ||
-						((sources & InputDevice.SOURCE_JOYSTICK) == InputDevice.SOURCE_JOYSTICK)) {
-					id = mJoysticksDevices.size();
-
-					Joystick joy = new Joystick();
-					joy.device_id = deviceId;
-					joy.name = device.getName();
-					joy.axes = new ArrayList<InputDevice.MotionRange>();
-					joy.hats = new ArrayList<InputDevice.MotionRange>();
-
-					List<InputDevice.MotionRange> ranges = device.getMotionRanges();
-					Collections.sort(ranges, new RangeComparator());
-
-					for (InputDevice.MotionRange range : ranges) {
-						if (range.getAxis() == MotionEvent.AXIS_HAT_X || range.getAxis() == MotionEvent.AXIS_HAT_Y) {
-							joy.hats.add(range);
-						} else {
-							joy.axes.add(range);
-						}
-					}
 
-					mJoysticksDevices.add(joy);
+		if (mJoystickIds.containsKey(deviceId)) {
+			return;
+		}
+
+		InputDevice device = mInputManager.getInputDevice(deviceId);
+		//device can be null if deviceId is not found
+		if (device == null) {
+			return;
+		}
+
+		int sources = device.getSources();
+
+		// Device may not be a joystick or gamepad
+		if (!(((sources & InputDevice.SOURCE_GAMEPAD) == InputDevice.SOURCE_GAMEPAD) ||
+					((sources & InputDevice.SOURCE_JOYSTICK) == InputDevice.SOURCE_JOYSTICK))) {
+			return;
+		}
+
+		// Assign first available number. Re-use numbers where possible.
+		final int id = assignJoystickIdNumber(deviceId);
+
+		final Joystick joy = new Joystick();
+		joy.device_id = deviceId;
+		joy.name = device.getName();
+
+		//Helps with creating new joypad mappings.
+		Log.i(tag, "=== New Input Device: " + joy.name);
 
-					final int device_id = id;
-					final String name = joy.name;
-					queueEvent(new Runnable() {
-						@Override
-						public void run() {
-							GodotLib.joyconnectionchanged(device_id, true, name);
-						}
-					});
+		//Collections.sort(ranges, new RangeComparator());
+		Set<Integer> already = new HashSet<Integer>();
+		for (InputDevice.MotionRange range : device.getMotionRanges()) {
+			boolean isJoystick = range.isFromSource(InputDevice.SOURCE_JOYSTICK);
+			boolean isGamepad = range.isFromSource(InputDevice.SOURCE_GAMEPAD);
+			//Log.i(tag, "axis: "+range.getAxis()+ ", isJoystick: "+isJoystick+", isGamepad: "+isGamepad);
+			if (!isJoystick && !isGamepad) {
+				continue;
+			}
+			final int axis = range.getAxis();
+			if (axis == MotionEvent.AXIS_HAT_X || axis == MotionEvent.AXIS_HAT_Y) {
+				joy.hasAxisHat = true;
+			} else {
+				if (!already.contains(axis)) {
+					already.add(axis);
+					joy.axes.add(axis);
+				} else {
+					Log.w(tag, " - DUPLICATE AXIS VALUE IN LIST: " + axis);
 				}
 			}
 		}
+		Collections.sort(joy.axes);
+		for (int idx = 0; idx < joy.axes.size(); idx++) {
+			//Helps with creating new joypad mappings.
+			Log.i(tag, " - Mapping Android axis " + joy.axes.get(idx) + " to Godot axis " + idx);
```

Comment:
Can this be moved in the previous `for` loop instead.

Reply:
Nope. The axes are populated using the previous for loop.